utm_tracker/models.py

```diff
@@ -1,8 +1,30 @@
+from __future__ import annotations
+
 from django.conf import settings
 from django.db import models
+from django.http import HttpRequest
 from django.utils import timezone
 
 
+class LeadSourceManager(models.Manager):
+    def create_from_request(self, request: HttpRequest) -> LeadSource:
+        """
+        Persist a LeadSource from an inbound HTTP Request.
+
+        This method 'pop's the values from the Django Session deliberately, so
+        that when a LeadSource is persisted - the session is cleared of the values
+        so that it doesn't persist a new LeadSource on every request.
+        """
+        LeadSource.objects.create(
+            user=request.user,
+            medium=request.session.pop("utm_medium"),
```

Comment:
Will this blow up on `AnonymousUser` - I'm not sure how Django handles that when saving?

Reply:
It will, but it's not going to be called like that - I can see from the middleware.

Reply:
It would blow up - but the middleware indeed doesn't call it like that.

I will raise a custom ValueError here, to give better information.

Reply:
Actually, this won't even work with the case I need to use it for as I'd need access to the actual `User` I wanted the LeadSource persisted against.

I'll rethink later.

Reply:
Updated: now passes `user` in explicitly, and explains why we don't just pluck `request.user` off the inbound request (because sometimes we want to assign a LeadSource to a known user from an unauthenticated request).